frontend/src/Exchange.tsx

```diff
@@ -124,10 +125,11 @@ class Exchange extends React.Component<IExchangeProps, IExchangeState> {
           <Trump trump={this.props.state.trump} />
           {this.props.state.game_mode !== 'Tractor' ? (
             <div>
-              <Friends gameMode={this.props.state.game_mode} />
+              <Friends gameMode={this.props.state.game_mode} showPlayed={false} />
               {this.state.friends.map((friend, idx) => {
                 const onChange = (x: IFriend) => {
                   const new_friends = [...this.state.friends];
+                  x.initial_skip = x.skip;
```

Comment:
Let's only set `initial_skip` in the frontend (i.e. not `x.skip`), and in the backend set `skip = initial_skip`

https://github.com/rbtying/shengji/blob/d126f0b8b5e8dc31ec966ec641f0edcecb3f0834/core/src/game_state.rs#L1068

That way, the backend is protected against someone derping with the websocket (which people do...)

Reply:
`friend` is a `&` reference, so the data it refers to cannot be written

Reply:
My suggestion would be something like:

```
pub struct FriendSelection {
  card: Card,
  initial_skip: usize,
}

pub fn set_friends(iter: impl IntoIterator<Item = FriendSelection>) -> ... {
  // ...
  let friends = iter.into_iter().map(|fs| Friend {
    card: fs.card,
    initial_skip: fs.initial_skip,
    skip: fs.initial_skip,
    player_id: None,
  });
}

```